olca-app/src/org/openlca/app/db/tables/CurrencyTable.java

```diff
@@ -0,0 +1,132 @@
+import org.openlca.app.viewers.tables.Tables;
+import org.openlca.core.model.Currency;
+import org.openlca.core.model.ModelType;
+
+public class CurrencyTable extends SimpleFormEditor {
+
+	private List<Currency> currencies;
+
+	public static void show() {
+		if (Database.get() == null) {
+			MsgBox.info(M.NoDatabaseOpened, M.NeedOpenDatabase);
+			return;
+		}
+		var id = "DbCurrencyTable";
+		Editors.open(new SimpleEditorInput(id, M.Parameters), id);
+	}
+
+	@Override
+	public void init(IEditorSite site, IEditorInput input)
+		throws PartInitException {
+		try {
+			currencies = Database.get().getAll(Currency.class);
+		} catch (Exception e) {
+			ErrorReporter.on("failed to load currencies", e);
+		}
+		super.init(site, input);
+	}
+
+	@Override
+	protected FormPage getPage() {
+		return new Page(this);
+	}
+
+	private static class Page extends FormPage {
+
+		private final List<Currency> currencies;
+
+		Page(CurrencyTable table) {
+			super(table, "DbCurrencyTable", M.Currencies);
+			currencies = table.currencies != null
+				? table.currencies
+				: Collections.emptyList();
+		}
+
+		@Override
+		protected void createFormContent(IManagedForm mform) {
+			var form = UI.formHeader(mform, M.Currencies);
+			var tk = mform.getToolkit();
+			var body = UI.formBody(form, tk);
+
+			var filterComp = tk.createComposite(body);
+			UI.gridLayout(filterComp, 2);
+			UI.gridData(filterComp, true, false);
+			var filter = UI.formText(filterComp, tk, M.Filter);
+
+			var table = Tables.createViewer(body,
+				M.Name,
+				M.Code,
+				"Exchange rate",
+				"ID");
+			Tables.bindColumnWidths(table, 0.3, 0.1, 0.2, 0.4);
+
+			var label = new Label();
+			table.setLabelProvider(label);
+			Viewers.sortByLabels(table, label, 0, 1, 2, 3);
+			table.setInput(currencies);
+			TextFilter.on(table, filter);
+			Actions.forRootEntities(table);
+		}
+	}
+
+	private static class Label extends LabelProvider
+		implements ITableLabelProvider {
+
+		@Override
+		public Image getColumnImage(Object obj, int col) {
+			if (!(obj instanceof Currency))
+				return null;
+			return switch (col) {
+				case 0 -> Images.get(ModelType.CURRENCY);
+				default -> null;
+			};
+		}
+
+		@Override
+		public String getColumnText(Object obj, int col) {
+			if (!(obj instanceof Currency currency))
+				return null;
+			return switch (col) {
+				case 0 -> currency.name;
+				case 1 -> currency.code;
+				case 2 -> getExchangeRate(currency, currency.referenceCurrency);
+				case 3 -> currency.refId;
+				default -> null;
+			};
+		}
+
+		// TODO (francois) Use a currency utility method as it is a "repetition" of
+		// TODO getExchangeRate in the currency editor.
+		private String getExchangeRate(Currency currency, Currency referenceCurrency) {
+			String s = "1 " + currency.code + " = ";
+			double f = currency.conversionFactor / referenceCurrency.conversionFactor;
+			f = Math.round(1000 * f) / 1000.0;
+			s += f + " " + referenceCurrency.code;
+			return s;
+		}
```

Comment:
I have a question here: I did not find a currency utility class in `olca-app`. If I want to gather this method and this [one](https://github.com/GreenDelta/olca-app/blob/b2154aaf3f2c8a219a0eb9aa56a4d71bd05ef549/olca-app/src/org/openlca/app/editors/currencies/CurrencyTable.java#L94). Should I:
- create a new currency specific utility class shared by `db.tables` and `db.editors`;
- add a method to `core.model.currency`.

Or maybe keep it separated.

Reply:
In this case it is just fine as it is. No need to extract this into a utility method.